# CBC Tutorials with Clarke :book:

Welcome to the CBC tutorials repository! :smiley:

This is where you can find a variety of helpful tips and tricks for phylogenetic analyses, and just general neat things to do in R. Please email me at <vsteenderen@gmail.com> if you have any questions or suggestions, or if you have some useful tips that you would like to share!


### :pushpin: **Each tutorial folder contains an instruction file with the extension .md.** 


## For phylogenetic tutorials, you need the following software:

* Chromas (free version 2.6.6) http://technelysium.com.au/wp/chromas/

* AliView http://ormbunkar.se/aliview/ or https://github.com/AliView/AliView

BioEdit is also widely used, but I prefer AliView as it can be used on multiple operating systems https://bioedit.software.informer.com/7.2/

* Notepad++ https://notepad-plus-plus.org/downloads/

* R https://cran.r-project.org/bin/windows/base/ 

* RStudio https://rstudio.com/products/rstudio/download/#download

* FigTree https://github.com/rambaut/figtree/releases

* Tracer https://github.com/beast-dev/tracer/releases

* PopART http://popart.otago.ac.nz/downloads.shtml

We'll go through using Inkscape later, for graphical editing https://inkscape.org/

## Useful online servers:

MAFFT (for multiple sequence alignment) https://mafft.cbrc.jp/alignment/software/

CIPRES (for running various analyses on a server) http://www.phylo.org/

EMBOSS seqret (for converting between file types, e.g. FASTA <-> NEXUS) https://www.ebi.ac.uk/Tools/sfc/emboss_seqret/

## YouTube tutorials:
### Make your own R package:
PART1: https://www.youtube.com/watch?v=pMq42QT8AbE&t=7s (Basic setup)

PART2: https://www.youtube.com/watch?v=1woJqeZCuog&t=34s (Adding a vignette)

PART3: https://www.youtube.com/watch?v=QsNAhukSw1g&t=59s (Embedding example data into the package)

### A guide to the BinMat R package for fragment analysis data

PART1: https://www.youtube.com/watch?v=RTHf7Naqr10

PART 2: https://www.youtube.com/watch?v=jk4uBzVgqw8
